pkg/background/gocraft.go

```diff
@@ -0,0 +1,202 @@
+		"queue":     queue,
+		"arguments": arguments,
+	}
+
+	crumbs.Debug(span.Context(), "Enqueueing job for background processing", map[string]interface{}{
+		"queue":     queue,
+		"arguments": arguments,
+	})
+
+	g.log.WithField("queue", queue).Debug("enqueuing job to be run")
+
+	encodedArguments, err := g.marshal(arguments)
+	if err = errors.Wrap(err, "failed to marshal arguments"); err != nil {
+		return err
+	}
+
+	_, err = g.enqueuer.EnqueueUnique(queue, map[string]interface{}{
+		"args": hex.EncodeToString(encodedArguments),
+	})
+	if err = errors.Wrap(err, "failed to enqueue job for gocraft/work"); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type GoCraftWorkJobProcessor struct {
+	state         uint32
+	configuration config.BackgroundJobs
+	log           *logrus.Entry
+	queue         *work.WorkerPool
+	enqueuer      JobEnqueuer
+	marshal       JobMarshaller
+}
+
+func NewGoCraftWorkJobProcessor(
+	log *logrus.Entry,
+	configuration config.BackgroundJobs,
+	redisPool *redis.Pool,
+	enqueuer JobEnqueuer,
+) *GoCraftWorkJobProcessor {
+	return &GoCraftWorkJobProcessor{
+		configuration: configuration,
+		log:           log,
+		queue:         work.NewWorkerPool(struct{}{}, 4, GoCraftWorkNamespace, redisPool),
+		enqueuer:      enqueuer,
+		marshal:       DefaultJobMarshaller,
+	}
+}
+
+func (g *GoCraftWorkJobProcessor) Start() error {
+	if !atomic.CompareAndSwapUint32(&g.state, 0, 1) {
+		return errors.New("gocraft/work job processor is either already started, or is in an invalid state")
+	}
+
+	g.queue.Start()
+	return nil
+}
+
+func (g *GoCraftWorkJobProcessor) Close() error {
+	if !atomic.CompareAndSwapUint32(&g.state, 1, 2) {
+		return errors.New("gocraft/work job processor is either already closed, or is in an invalid state")
+	}
+
+	doneChannel := make(chan struct{})
+	timeout := time.NewTimer(30 * time.Second)
+
+	// queue.Drain can technically run forever. So run it in a go-routine and defer the done signal for when it finishes
+	// if it ever finishes. If it does not then it will be killed if we timeout anyway.
+	go func() {
+		defer func() {
+			doneChannel <- struct{}{}
+		}()
+		g.queue.Drain()
+	}()
+
+	defer func() {
+		g.queue.Stop()
+	}()
+	select {
+	case <-doneChannel:
+		return nil
+	case <-timeout.C:
+		return errors.New("timeout while draining gocraft/work queue")
+	}
+}
+
+func (g *GoCraftWorkJobProcessor) RegisterJob(ctx context.Context, handler JobHandler) error {
+	log := g.log.WithField("job", handler.QueueName())
+	log.Trace("registering job handler")
+	g.queue.Job(handler.QueueName(), func(job *work.Job) (err error) {
+		highContext := sentry.SetHubOnContext(context.Background(), sentry.CurrentHub().Clone())
+		span := sentry.StartSpan(highContext, "queue.process", sentry.TransactionName(handler.QueueName()))
+		jobLog := log.WithContext(span.Context())
+		hub := sentry.GetHubFromContext(span.Context())
+		hub.ConfigureScope(func(scope *sentry.Scope) {
+			scope.SetTag("queue", handler.QueueName())
+		})
```

Comment:
Would probably be a good idea to add comments about what all is being done here. This is a bit messy.